### PR TITLE
Add AzureBagVerifier

### PR DESCRIPTION
**Pull request by alicefuzier**

### CODE REVIEW

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/bagit/services/azure/AzureBagReaderTest.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import uk.ac.wellcome.storage.store.azure.{AzureStreamStore, AzureTypedStore}
+
+class AzureBagReaderTest extends BagReaderTestCases[
+  Unit,
+  Container,
+  AzureBlobLocation,
+  AzureBlobLocationPrefix
+]
+  with AzureBagBuilder {
+  override def asLocation(root: AzureBlobLocationPrefix, path: String): AzureBlobLocation = root.asLocation(path)
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/platform/archive/common/bagit/services/azure/AzureBagReaderTest.scala:18]: Note to self: I think we guarantee an `asLocation` method on `Prefix[Location]` in the newest versions of scala-libs.

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/azure/AzureBagBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  with AzureFixtures{
+  override def createBagRoot(space: StorageSpace, externalIdentifier: ExternalIdentifier, version: BagVersion)(implicit container: Container): AzureBlobLocationPrefix = AzureBlobLocationPrefix(container.name, createBagRootPath(space, externalIdentifier, version))
+
+  override def createBagLocation(bagRoot: AzureBlobLocationPrefix, path: String): AzureBlobLocation = AzureBlobLocation(bagRoot.container, path)
+
+  override protected def buildFetchEntryLine(entry: PayloadEntry)(implicit container: Container): String = {
+    val displaySize =
+      if (Random.nextBoolean()) entry.contents.getBytes.length.toString else "-"
+
+    s"""https://myaccount.blob.core.windows.net/${container.name}/${entry.path} $displaySize ${entry.bagPath}"""
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/azure/AzureBagBuilder.scala:22]: Hey, somebody’s actually gone and done the homework :P

I just made up my own URL scheme when I tried this.

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/BagVerifier.scala [hunk 3]
```diff
@@ -136,7 +136,7 @@ trait BagVerifier[BagContext <: BagVerifyContext[BagPrefix], BagLocation <: Loca
       _ <- verifyFetchPrefixes(
         fetch = bag.fetch,
         root = S3ObjectLocationPrefix(
-          bucket = primaryBucket,
+          bucket = namespace,
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/BagVerifier.scala:139]: The use of "bucket" rather than "namespace" is deliberate here – the primary copy of a bag is always in an S3 bucket, not an Azure container or anywhere else.

File: bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  override def withVerifier[R](container: Container)(
+    testWith: TestWith[ReplicatedBagVerifier[
+      S3ObjectLocation,
+      S3ObjectLocationPrefix,
+      AzureBlobLocation,
+      AzureBlobLocationPrefix
+    ], R]
+  )(implicit typedStore: TypedStore[AzureBlobLocation, String]): R =
+    testWith(
+      new AzureReplicatedBagVerifier(namespace = container.name)
```
Comment (alexwlchan) [inline bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala:97]: I would expect you to be passing a bucket here, for the "primary" copy of a bag, so I suspect something isn't quite right.

File: bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      Container
+    ]
+    with S3Fixtures
+    with AzureFixtures {
+
+  val s3TypedStore = S3TypedStore[String]
+  val azureTypedStore = new AzureTypedStore[String]()(
+    implicitly[Codec[String]],
+    new AzureStreamStore()
+  )
```
Comment (alexwlchan) [inline bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala:40]: I think AzureTypedStore has the same constructor as S3TypedStore:

```suggestion
  val azureTypedStore = AzureTypedStore[String]
```

Comment (alicefuzier) [reply bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala:40]: not in the version of scala-libs we use right now. Should be added definitely, but I will focus on this PR first

File: bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala (hunk 4 above)
Comment (alicefuzier) [reply bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala:97]: ah thanks, will change

File: bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala (hunk 5 above)
Comment (alexwlchan) [reply bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala]: Oops, my bad. I have distinct memories of writing such a constructor; but it must have got lost in a branch somewhere.

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/fixity/FixityChecker.scala [hunk 6]
```diff
@@ -81,7 +82,7 @@ trait FixityChecker[BagLocation <: Location] extends Logging {
   private def parseLocation(
     expectedFileFixity: ExpectedFileFixity
   ): Either[FileFixityCouldNotRead[BagLocation], BagLocation] =
-    locate(expectedFileFixity.uri) match {
+    expectedFileFixity.uri.locate match {
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/fixity/FixityChecker.scala:85]: 👍

Good spot (I think the `locate()` method came in when I was disambiguating ObjectLocation, and I forgot to revert it).

How hard would it be to pull out into a separate patch? That would be mergeable immediately and make this diff simpler.

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/models/BagVerifyContext.scala [hunk 7]
```diff
@@ -11,9 +12,8 @@ case class StandaloneBagVerifyContext[BagPrefix <: Prefix[_ <: Location]](
 ) extends BagVerifyContext[BagPrefix]
 
 case class ReplicatedBagVerifyContext[
-  SrcBagPrefix <: Prefix[_ <: Location],
   ReplicaBagPrefix <: Prefix[_ <: Location]
-](srcRoot: SrcBagPrefix, replicaRoot: ReplicaBagPrefix)
+](srcRoot: S3ObjectLocationPrefix, replicaRoot: ReplicaBagPrefix)
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/models/BagVerifyContext.scala:16]: 👍

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/BagVerifier.scala [hunk 8]
```diff
@@ -75,7 +59,9 @@ trait BagVerifier[BagContext <: BagVerifyContext[BagPrefix], BagLocation <: Loca
     bag: Bag
   ): Either[BagVerifierError, Unit]
 
-  def createPrefix(path: String): BagPrefix
+  // What bucket is storing the "primary" copy of a bag?  This is the bucket
+  // that should be referred to in the fetch.txt URIs.
+  val bucket: String
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/BagVerifier.scala:64]: nit: "primaryBucket"? So we don't get it confused with other buckets we might be using.

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import uk.ac.wellcome.storage.listing.Listing
+import uk.ac.wellcome.storage.listing.azure.AzureBlobLocationListing
+import uk.ac.wellcome.storage.s3.S3ObjectLocation
+import uk.ac.wellcome.storage.store.StreamStore
+import uk.ac.wellcome.storage.store.azure.AzureStreamStore
+import uk.ac.wellcome.storage.store.s3.S3StreamStore
+
+class AzureReplicatedBagVerifier(val bucket: String)(
+  implicit val s3Client: AmazonS3,
+  implicit val blobClient: BlobServiceClient
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala:21]: nit: is there a reason these are both `val`?

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/BagVerifier.scala [hunk 10]
```diff
@@ -12,38 +12,22 @@ import uk.ac.wellcome.platform.archive.common.bagit.services.BagReader
...
-    ] {
-  override def verifyReplicatedBag(
-    root: StandaloneBagVerifyContext[BagPrefix],
-    space: StorageSpace,
-    externalIdentifier: ExternalIdentifier,
-    bag: Bag
-  ): Either[BagVerifierError, Unit] = Right(())
-}
-
 trait ReplicatedBagVerifier[
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/BagVerifier.scala:20]: minor: is it worth splitting ReplicatedBagVerifier out into its own file, since S3StandaloneBagVerifier now lives elsewhere?

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala (hunk 9 above)
Comment (alexwlchan) [reply bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala:21]: So now I look at this, I wonder – maybe you expect an implicit `AmazonS3` in ReplicatedBagVerifier, and declare the `srcStreamStore` there?

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/s3/S3BagVerifier.scala [hunk 11]
```diff
@@ -54,26 +55,32 @@ trait S3BagVerifier[B <: BagVerifyContext[S3ObjectLocationPrefix]]
...
+      S3ObjectLocation,
+      S3ObjectLocationPrefix
+    ]
+    with S3BagVerifier[StandaloneBagVerifyContext[S3ObjectLocationPrefix]] {
+  override def verifyReplicatedBag(
+    root: StandaloneBagVerifyContext[S3ObjectLocationPrefix],
+    space: StorageSpace,
+    externalIdentifier: ExternalIdentifier,
+    bag: Bag
+  ): Either[BagVerifierError, Unit] = Right(())
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/s3/S3BagVerifier.scala:71]: Can we delete this now?

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/storage/azure/AzureResolvable.scala [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+class AzureResolvable(implicit blobServiceClient: BlobServiceClient)
+    extends Resolvable[AzureBlobLocation] {
+  override def resolve(t: AzureBlobLocation): URI = {
+    new URI(
+      blobServiceClient
+        .getBlobContainerClient(t.container)
+        .getBlobClient(t.name)
+        .getBlobUrl
+    )
```
Comment (alexwlchan) [inline bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/storage/azure/AzureResolvable.scala:17]: nit: let's use `location` for the variable name

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala (hunk 9 above)
Comment (alexwlchan) [reply bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala:21]: Aha, I see now, I hadn't read all the way down – so can you delete the `srcStreamStore` here?

File: bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/ReplicatedBagVerifierTestCases.scala [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+          result shouldBe a[IngestFailed[_]]
+          result.summary shouldBe a[VerificationIncompleteSummary]
+
+          result.maybeUserFacingMessage shouldNot be(defined)
+      }
+    }
+  }
+
+  it("fails a bag if it cannot read the original bag") {
```
Comment (alexwlchan) [inline bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/ReplicatedBagVerifierTestCases.scala:66]: It's not clear to me why this test fails – is it because the original bag never gets written?

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/BagBuilder.scala [hunk 14]
```diff
@@ -47,7 +47,8 @@ trait BagBuilder[BagLocation <: Location, BagPrefix <: Prefix[BagLocation], Name
     space: StorageSpace = createStorageSpace,
     externalIdentifier: ExternalIdentifier = createExternalIdentifier,
     version: BagVersion = BagVersion(randomInt(from = 2, to = 10)),
-    payloadFileCount: Int = randomInt(from = 5, to = 50)
+    payloadFileCount: Int = randomInt(from = 5, to = 50),
+    bucketName: String = randomAlphanumeric
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/BagBuilder.scala:51]:
```suggestion
    bucketName: String = createBucket
```

Not all random alphanumeric strings are valid bucket names, and you can get flaky test failures depending on whether or not the random string is valid.

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/BagBuilder.scala [hunk 15]
```diff
@@ -136,21 +137,30 @@ trait BagBuilder[BagLocation <: Location, BagPrefix <: Prefix[BagLocation], Name
...
 
   protected def buildFetchEntryLine(
+    bucketName: String,
     entry: PayloadEntry
-  )(implicit namespace: Namespace): String
+  ): String = {
+    val displaySize =
+      if (Random.nextBoolean()) entry.contents.getBytes.length.toString else "-"
+
+    s"""s3://${bucketName}/${entry.path} $displaySize ${entry.bagPath}"""
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/BagBuilder.scala:162]: 👍

Smart

File: bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/ReplicatedBagVerifierTestCases.scala (hunk 13 above)
Comment (alicefuzier) [reply bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/ReplicatedBagVerifierTestCases.scala:66]: I think something is not set up correctly right now in the tests. None of them can find the bag where is supposed to be

File: common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/BagBuilder.scala (hunk 14 above)
Comment (alicefuzier) [reply common/src/test/scala/uk/ac/wellcome/platform/archive/common/fixtures/BagBuilder.scala:51]: thanks

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/fixity/FixityChecker.scala (hunk 6 above)
Comment (alicefuzier) [reply bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/fixity/FixityChecker.scala:85]: yes, unfortunately the refactoring commits are a bit entangled... the last commit is entirely refactoring, but relies on other refactoring I did before... I think it might be a bit complicated now

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala (hunk 9 above)
Comment (alicefuzier) [reply bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala:21]: Good spot, they don't need to be both val

Comment (alexwlchan) [reply bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala:21]: I think maybe it does need to be val after all, but the compiler will tell you for sure.

Comment (alicefuzier) [reply bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifier.scala:21]: the s3Client does, but the blobClient doesn't. Anyway I've also removed srcStreamStore, thanks

File: bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala (hunk 5 above)
Comment (alicefuzier) [reply bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala]: I updated the version of scala libs in this branch. Thanks @alexwlchan for fixing that

Comment (alexwlchan) [reply bag_verifier/src/test/scala/uk/ac/wellcome/platform/archive/bagverifier/services/azure/AzureReplicatedBagVerifierTests.scala]: You may regret doing that; said library is a bit broken, sorry: https://github.com/wellcomecollection/platform/issues/4718

**Comment by alicefuzier** [comment]
Yep, I think I can isolate at least adding the new Azure classes. Also, i missed that you already had a PR to upgrade scalalibs, thanks. I wasn't going to do it in this PR but then needed a `AzureTypedStore` for the `AzureSizeChecker` and didn't want to create one cause it's annoying. I figured just upgrading libs would be easier

File: bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/s3/S3BagVerifier.scala (hunk 11 above)
Comment (alicefuzier) [reply bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/services/s3/S3BagVerifier.scala]: I don't think so, It needs to be implemented as a no-op thing somewhere. Either here or in the base `BagVerifier` trait. I prefer here cause it means the `BagVerifier` makes fewer assumptions but ..meh

**Comment by alicefuzier** [comment]
Closing as this has been merged as a sequence of PRs kindly split by @alexwlchan from this one.
